Rflink 433Mhz gateway platform and components

aequitas opened the pull request:
**Description:** The change is intended to add support for the Rflink gateway devices (http://www.nemcon.nl/blog2/ & https://www.nodo-shop.nl/nl/21-rflink-gateway) which allow receiving and sending of 433Mhz packets much like rfxtrx. It builds on the rflink python package (https://pypi.python.org/pypi/rflink) which was initiated for this cause.

At the moment this is a very rough (but working) sketch with a lot of debugging. Expect much of the code to be changed, move around or migrate towards the rflink package to be available for a broader audience.

**Feel free to leave feedback or feature requests.**

I try to model this close to the rfxtrx platform but since this platform is developed asyncio first not a lot can be shared (yet).

**Please help by testing this out with your own Rflink device and providing feedback.**

What works at the moment:
- [x] Receiving and sending of on/off switch events to control lights.
- [x] Automatically adding discovered devices (but not stored yet)
- [x] Temparature & humidity sensors

Todos:
- [x] add tests
- [x] fix linting
- [x] cleanup
- [ ] doorbell support
- [ ] covers support? 
- [x] persistent devices from configuration.yaml
- [ ] assumed state configuration? (per device?)
- [x] temperature unit per device configuration
- [x] (debug) logging through log framework
- [x] shared known devices (light can be configured as switch)
- [x] component.rflink spawns devices? or event to create platform device?
- [ ] use service pattern instead of direct bus events? 
- [ ] rflink unsupported devices debug enable
- [ ] leave last box unchecked
- [x] config validation
- [ ] loading previous state from recorder

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#1715

**Example entry for `configuration.yaml` (if applicable):**
```yaml
rflink:
  port: /dev/serial/by-id/usb-Arduino__www.arduino.cc__0042_55632313338351B08080-if00
  # send simultaneous commands as quickly as possible (may cause commands to fail)
  # wait_for_ack: false
  # ignore_device:
  #   - digitech_*
 
light:
  - platform: rflink
    # adds new devices to this group if not configured below
    new_devices_group: New Rflink Lights
    devices:
      newkaku_00000001_a:
        name: bedroom
      newkaku_00000001_b:
        name: kitchen
        aliasses:
          - kaku_000000_3

sensor:
  - platform: rflink
    # adds new devices to this group if not configured below
    new_devices_group: New Rflink Sensors
    devices:
      digitech_002f:
        # a single rflink packet can contain multiple values (eg: temperature and humidity)
        # setting value_key allows to specify which key should be used from the packet
        value_key: temperature

switch:
  - platform: rflink
    devices:
      newkaku_00000001_1:
        name: ventilator
        icon: mdi:fan

```

If you want to test this on for example your laptop/desktop without having to modify your existing HASS server but you don't want to have your Rflink hooked to the serial port of your laptop/desktop, this module support connecting through TCP as well. On the host connected to the Rflink install `socat` and run:

```bash
socat -ddd /dev/ttyACM0,b57600 TCP-LISTEN:1234,reuseaddr
```

This will expose the serial line on a tcp connection.

Then configure the component like this:
```yaml
rflink:
  host: hostname
  port: 1234

light:
  - platform: rflink
```

aequitas committed:
Initial sketches of rflink component.

aequitas committed:
Add requirement.

aequitas committed:
Properly load configuration.

aequitas committed:
Bump rflink for graceful parse errors and protocol callback.

aequitas committed:
Cleanup, documentation and linting.

aequitas committed:
More documentation, first sensor implementation (temp & hum).

aequitas committed:
Add brightness/dim support for newkaku protocol.

aequitas committed:
Use separate class for dimmables.

aequitas committed:
Make sure non-dimmable newkaku devices are turned on.

aequitas committed:
Move some code around, add switches. Support loading from config.

aequitas commented:
Support for loading devices from configuration added.

aequitas committed:
Fix bug in ignoring devices.

aequitas committed:
Fix initial state assumption.

aequitas committed:
Improve reliability on invalid conditions.

aequitas committed:
Allow configuration of group for new devices.

aequitas committed:
Sensor icons.

aequitas committed:
Fix parsing negative numbers.

aequitas committed:
Correct icon.

aequitas committed:
Allow sending commands serial.

aequitas committed:
Pluralize.

balloob commented:
Please change title and leave a note when ready for review 👍

aequitas committed:
Allow adding sensors from config.

aequitas committed:
Fix ignoring devices and bugs in previous commit.

aequitas committed:
Share know devices so devices from configuration don't get added as lights.

aequitas committed:
Lookup unit from value_key.

Danielhiversen commented:
Any progress on this?

aequitas commented:
@Danielhiversen Currently working on refactoring the Python module. Hope to work those changes into the component this weekend to simplify the HA implementation a little.

mjj4791 commented:
@aequitas 
I was working on an rflink component as well; but found your implementation cleaner. So I started evaluating your implementation. Overall, not disappointed at all!

If you are interested, I used your first beta and added:
- [x] explicit config checking
- enhanced reading sensor config (current implementation only allows for 1 value_type per sensor in configfile)
   - allow for multiple value_types per sensordevice
   - [x] configure unit of measurement per value_type (with fallback to default unit)
- added conversions in rflink library for all value types rflink supports

Two things to add to your todo list:
- [ ] (re)connecting to rflink on separate reconnect loop (reconnect when disconnected & allow HA to start when rflink is not found (yet))
- [ ] add signal-repetitions config to send out commands more than once (sometimes lights are not switched due to distance / other devices transmitting)

I have attached my modifications to your code in a zip file (based on rflink 0.0.7 library)
[rflink.zip](https://github.com/home-assistant/home-assistant/files/659399/rflink.zip)

pepeEL commented:
Hi
But how i can add configuration for my sensor temp from weather station ? In Domoticz it will be automatically. Now how i can to tghis step by step in HA ?

aequitas commented:
@mjj4791 cool, thanks for contributing, I will check it out soon.

aequitas commented:
@pepeEL it will automatically detect and add any new sensor it finds. Currently not all sensors are supported, this should be fixed soon.

pepeEL commented:
But can you describe step by step how add RFLink Gateway to Home Assistan and then how add devices like sensor temp, like blinds somfy, like switch anslut(from Jula Shop) or sensor/switch contact ? Please i am begginer and i want to go to working HA but now i use Domoticz and Domoticz is no good.

Danielhiversen commented:
@pepeEL This is work in progress, and it is currently not added to the official version of HA

pepeEL commented:
ok i am waiting for add officialy with good easy manual how configure it and add devices.

aequitas committed:
Remove debug.

aequitas commented:
@mjj4791 I added your changes to rflink: https://github.com/aequitas/python-rflink/commit/42e4380a1acf709f51c3800e1abe3910486d460a

Refactored them a little to match the style and added the tests.

For the changes to home-assistant, could you please create a pullrequest on this branch? It will be easier for me to maintain and give feedback.

boschr commented:
Keep up the good work! When RFLink is implemented I will try HASS instead of Domoticz.

pepeEL commented:
I also wait for working RFLink and also start use HA instead Domoticz. But now i search how use Razberry Zwave. How install and how pair with devices.

kabturek commented:
Thanks for work, i've tested it in HA and it discovered my switches (and my neighbors too ;-) 
I have some problems with OFF state beeing ON but that is a problem with rflink too afaik (OFF turns ON my light). 
Good work!

kabturek commented:
yeah so i've found that i have to:
`10;TRISTATEINVERT=ON;`
to rflink, will test at home

aequitas commented:
@kabturek yea, thats a 'problem' I have to with my neighbors :)

Let me know what you find out and I can incorporate it into HA/rflink module.

aequitas committed:
Start implementing event protocol in place of packet protocol.

- Added first test suite for sensors.
- This currently breaks light and switch.

aequitas commented:
I am currently rewriting the component to the new rflink python module implementation. Currently functionality for switches/lights is broken. Sensors are now tested and should work.

aequitas committed:
Refactor switch component to fit new rflink changes. Add test suite.

aequitas committed:
Fix style.

mjj4791 commented:
aequitas

 I tried your commit from 24 dec.

Some feedback:
- If i switch a device using the remote control, I see the following exception in the log:
  This happens both by on and of switching...
  
 15:25:14 rflink.protocol: received data: 20;87;FA500;ID=0000
 15:25:14 rflink.protocol: received data: XXXX;SWITCH=01;CMD=ON;
 15:25:14 rflink.protocol: got packet: 20;87;FA500;ID=0000XXXX;SWITCH=01;CMD=ON;
 15:25:14 rflink.protocol: decoded packet: {'node': 'gateway', 'protocol': 'fa500', 'id': '0000XXXX', 'command': 'on', 'switch': '01'}
 15:25:14 rflink.protocol: got event: {'id': 'fa500_0000XXXX_01', 'command': 'on'}
 15:25:14 homeassistant.components.rflink: event type light
 15:25:14 homeassistant.core: Bus:Handling <Event rflink_switch_event_received[L]: event=id=fa500_0000XXXX_01, command=on>
 15:25:14 homeassistant.core: Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/lib/python3.4/asyncio/tasks.py", line 238, in _step
    result = next(coro)
  File "/srv/hass/lib/python3.4/site-packages/homeassistant/components/light/rflink.py", line 84, in add_new_device
    entity_type = entity_type_for_device_id(packet['protocol'])
KeyError: 'protocol'
 15:25:14 homeassistant.core: Bus:Handling <Event state_changed[L]: new_state=<state light.kerstverlichting=on; friendly_name=kerstverlichting, supported_features=0 @ 20T15:25:14.529090+01:00>, old_state=<state light.kerstverlichting=off; friendly_name=kerstverlichting, assumed_state=True, supported_features=0 @ 20T15:23:57.581300+01:00>, entity_id=light.kerstverlichting>
 15:25:14 homeassistant.core: Bus:Handling <Event state_changed[L]: new_state=<state group.all_lights=on; order=2, friendly_name=all lights, assumed_state=True, entity_id=('light.schuur', 'light.kerstverlichting'), hidden=True, auto=True @ 20T15:25:14.532219+01:00>, old_state=<state group.all_lights=off; order=2, friendly_name=all lights, assumed_state=True, entity_id=('light.schuur', 'light.kerstverlichting'), hidden=True, auto=True @ 20T15:23:57.586207+01:00>, entity_id=group.all_lights>


- If I switch from the HA gui; the lights are not switched at all....

 15:27:24 homeassistant.components.http: Serving /api/services/homeassistant/turn_on to 127.0.0.1 (auth: True)
 15:27:24 homeassistant.core: Bus:Handling <Event call_service[L]: domain=homeassistant, service=turn_on, service_data=entity_id=light.kerstverlichting, service_call_id=139858937591848-5>
 15:27:24 homeassistant.core: Bus:Handling <Event call_service[L]: domain=light, service=turn_on, service_data=entity_id=['light.kerstverlichting'], service_call_id=139858937591848-6>
 15:27:24 homeassistant.core: Bus:Handling <Event rflink_send_command[L]: command=on, entity_id=fa500_00002728_01>
 15:27:24 rflink.protocol: sending command: {'protocol': 'fa500', 'switch': '01', 'id': '00002728', 'command': 'on'}
 15:27:24 rflink.protocol: writing data: 10;fa500;00002728;01;on;\r\n
 15:27:24 rflink.protocol: waiting for acknowledgement
 15:27:24 homeassistant.core: Bus:Handling <Event service_executed[L]: service_call_id=139858937591848-6>
 15:27:24 homeassistant.core: Bus:Handling <Event service_executed[L]: service_call_id=139858937591848-5>
...
 15:27:29 rflink.protocol: acknowledge timeout
 15:27:30 homeassistant.core: Bus:Handling <Event state_changed[L]: new_state=<state sun.sun=above_horizon; next_setting=20T15:33:51+00:00, friendly_name=Sun, next_rising=2016-12-26T07:49:53+00:00, azimuth=217.68, elevation=6.42 @ 20T15:23:58.525636+01:00>, old_state=<state sun.sun=above_horizon; next_setting=20T15:33:51+00:00, friendly_name=Sun, next_rising=2016-12-26T07:49:53+00:00, azimuth=217.47, elevation=6.51 @ 20T15:23:58.525636+01:00>, entity_id=sun.sun>

Looks like no response is received from rflink (mine is remoet over IP); maybe something wrong with command termination? taht would explain why rflink is not reponding....

For now I revert back to my previous deployed version....

aequitas commented:
@mjj4791 thanks for reporting, like said above, it is currently being rewritten to fit refactoring of the rflink module. Some stuff is still broken until I find some lost hours to work on it.

aequitas committed:
Refactor and test lights. Bring coverage to 100%.

aequitas commented:
Test coverage is now complete and lights should be working again.

aequitas committed:
Use non-broken and production tested rflink module.

aequitas committed:
Update requirements.

aequitas committed:
Bump for logging.

balloob commented:
Is this still WIP?

aequitas commented:
@balloob no longer, was just preparing to remove the wip status :). 

In the current state it is functional and useful. There are a lot more features to be added, but nothing really blocking a release.

I will start working on the home-assistant.github.io documentation shortly.

balloob commented on homeassistant/components/light/rflink.py:
```diff
@@ -0,0 +1,136 @@
+
+
+def entity_type_for_device_id(device_id):
+    """Return entity class for procotol of a given device_id."""
```
Since you're writing async code, please indicate in the doc string if this method is safe to run in both async and non-async world. [like this](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/core.py#L368)

balloob commented on homeassistant/components/light/rflink.py:
```diff
@@ -0,0 +1,136 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+VALID_CONFIG_KEYS = [
```
Please replace this const and the associated check with a `PLATFORM_SCHEMA`

balloob commented on homeassistant/components/light/rflink.py:
```diff
@@ -0,0 +1,136 @@
+    devices = []
+    for device_id, config in domain_config.get('devices', {}).items():
+        # extract only valid keys from device configuration
+        kwargs = {k: v for k, v in config.items() if k in VALID_CONFIG_KEYS}
```
PLATFORM_SCHEMA

balloob commented on homeassistant/components/rflink.py:
```diff
@@ -0,0 +1,269 @@
+
+        # fire bus event for event type
+        if event_type in RFLINK_EVENT:
+            hass.bus.fire(RFLINK_EVENT[event_type], {ATTR_EVENT: event})
```
Instead of using the event bus, have platforms register with the main component via a class and distribute data via that.

aequitas replied:
I see this pattern used in many other components and I think its a nice way of decoupling. What is the reason it should not be used so I can make a better judgement in the future to apply this pattern.

balloob replied:
The pattern of having internal communication between a component and it's platform being done over the event bus is not something that we have in Home Assistant. If it sneaked in, it should be removed. There are definitely not "many components".

Only use the event bus for data that can be relevant to the user. Communication between a platform and the actual device is not relevant as the platform will process this information and write the relevant information to the state machine for user consumption.

aequitas replied:
Sorry, many was a bit of an overstatement indeed. Two components that do seem to use this pattern are:
- https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/zigbee.py#L128-L135
- https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/switch/pilight.py#L90-L91

Others just output to the bus for user to subscribe to but don't use it internally indeed, did not see that before, sorry. 

I will refactor to not use the bus for propagating events internally. 

I think I will keep firing some of the events to allow the user to subscribe to things like button presses instead of just state changes.

DavidLP replied:
> The pattern of having internal communication between a component and it's platform being done over the event bus is not something that we have in Home Assistant.

Good to know. I thought the concept was to have hub components for communication serialization with the "outside world" (hardware, libraries, ...) and hass components using this hub via the event bus.

balloob replied:
Yeah, we should probably document this better. (but we already have so much documentation that people don't read it… )

balloob commented on homeassistant/components/light/rflink.py:
```diff
@@ -0,0 +1,136 @@
+        entity_type = entity_type_for_device_id(event['id'])
+        entity_class = entity_class_for_type(entity_type)
+
+        if event['id'] not in rflink.KNOWN_DEVICE_IDS:
```
use a guard clause to maintain readability:

```python
if event['id'] in rflink.KNOWN_DEVICE_IDS:
    return

…
```

balloob commented on homeassistant/components/rflink.py:
```diff
@@ -0,0 +1,269 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+KNOWN_DEVICE_IDS = []
```
Do not use global data structures. Instead store data in dictionary `hass.data`

balloob commented on homeassistant/components/rflink.py:
```diff
@@ -0,0 +1,269 @@
+            _LOGGER.debug('unhandled event of type: %s', event_type)
+
+    # when connecting to tcp host instead of serial port (optional)
+    host = config[DOMAIN].get('host', None)
```
Add a CONFIG_SCHEMA to set defaults like `None`

balloob commented on homeassistant/components/rflink.py:
```diff
@@ -0,0 +1,269 @@
+        Does not wait for Rflink to confirm the packet has been sent.
+
+        """
+        protocol.send_command(
```
this is not a coroutine?

aequitas replied:
No, just a normal function, puts the command on the out buffer and returns, transport/protocol takes care of asynchronous write to the IO. Annotated this in the docstring for the function.

balloob commented on homeassistant/components/rflink.py:
```diff
@@ -0,0 +1,269 @@
+            event.data[ATTR_COMMAND],
+        )
+
+    if config.get('wait_for_ack', True):
```
Move default value to CONFIG_SCHEMA

balloob commented on homeassistant/components/sensor/rflink.py:
```diff
@@ -0,0 +1,121 @@
+"""Support for Rflink lights.
```
stale comment

balloob commented on homeassistant/components/sensor/rflink.py:
```diff
@@ -0,0 +1,121 @@
+    devices = []
+    for device_id, config in domain_config.get('devices', {}).items():
+        # extract only valid keys from device configuration
+        kwargs = {k: v for k, v in config.items() if k in VALID_CONFIG_KEYS}
```
Use PLATFORM_SCHEMA

balloob commented on homeassistant/components/sensor/rflink.py:
```diff
@@ -0,0 +1,121 @@
+        """Check if device is known, otherwise create device entity."""
+        event = event.data[rflink.ATTR_EVENT]
+        device_id = event['id']
+        if device_id not in rflink.KNOWN_DEVICE_IDS:
```
guard clause

balloob commented on homeassistant/components/switch/rflink.py:
```diff
@@ -0,0 +1,52 @@
+    devices = []
+    for device_id, config in domain_config.get('devices', {}).items():
+        # extract only valid keys from device configuration
+        kwargs = {k: v for k, v in config.items() if k in VALID_CONFIG_KEYS}
```
PLATFORM_SCHEMA

balloob commented on homeassistant/components/switch/rflink.py:
```diff
@@ -0,0 +1,52 @@
+    @property
+    def icon(self):
+        """Return the icon to use for device if any."""
+        return self._icon
```
I would expect the property that returns `self._icon` to live in the same class as that defines `self._icon`.

aequitas replied:
I removed custom icon support altogether from the component, as customizations already provide this feature. The only entity that still uses custom icons are sensors which can detect an appropriate icon based on the thing they represent.

aequitas commented:
@balloob thanks for the feedback. Will start working on it shortly.

aequitas committed:
Improve readability.

Genestealer commented:
Can I please ask how things are getting on with this? Really interested. :)

kabturek commented:
i am using an earlier version (before refactor) and its working great (only switches and some sensors from someones weather station). Will test the latest version now.

Great work @aequitas !

KirovAir commented:
I just wanted to say, impressive work @aequitas !
As soon as I receive my Rflink I'll start testing and contribute where possible.

aequitas commented:
Sorry for the delay in finishing this. Will start working on it real soon. I rebased it to latest release (0.36.1) so anyone wanting to use it already or test can try installing using:

    pip3 install https://github.com/aequitas/home-assistant/archive/rflink.zip --upgrade

Genestealer commented:
Brilliant I'm really looking forward to trying this out on my windows 10 based homeassistant with the NodeMCU serial wifi Bridge.

aequitas committed:
Do not use global variable but keep known device state in intended place.

aequitas committed:
Improve docs.

aequitas committed:
Make icon support generic.

aequitas committed:
Disable overriding icons in config, as it belongs in customization. Only keep custom icon for entities that are able to detect a icon based on the thing they represent (sensors in this case).

aequitas committed:
Implement configuration schema, overall refactor of magic values.

aequitas committed:
Fix bug in config/test wait_for_ack.

aequitas committed:
Small refactors.

aequitas commented:
Most of the open topics are tackled, just one open question. Code not tested yet in real life.

lmolent commented:
Hi, many thanks for you component @aequitas, but after applying last commits I see two regressions:

- [ ] icon: mdi:fan is no longer allowed in switch section
- [ ] UnicodeEncodeError: 'ascii' codec can't encode character '\xb0' in position 101: ordinal not in range(128)  - after each recieved RF packet

And I agree with @mjj4791, please add signal-repetitions option.

aequitas commented:
@lmolent thanks for the feedback.

The icon feature is not planned to come back, as that function is already provided by customizations: https://home-assistant.io/getting-started/customizing-devices/

For the UnicodeEncodeError, could you maybe install the Rflink python package manually `pip3 install rflink` and run the rflink command with debugging on (HA should be stopped or not listening on the same serial port): `rflink -vv`. See if this gives the same errors. If it does could you send me some of the debug logging.

lmolent commented:
Hi @aequitas, rflink -vv doesn't produce these errors:

`root@homeassistant:~# rflink -vv -p /dev/ttyUSB0
DEBUG:asyncio:Using selector: EpollSelector
DEBUG:rflink.protocol:connected
DEBUG:rflink.protocol:received data: 20;00;Nodo RadioFrequencyLink -
DEBUG:rflink.protocol:received data: RFLink Gateway V1.1 - R45;
DEBUG:rflink.protocol:got packet: 20;00;Nodo RadioFrequencyLink - RFLink Gateway V1.1 - R45;
DEBUG:rflink.protocol:decoded packet: {'version': '1.1', 'protocol': 'unknown', 'hardware': 'Nodo RadioFrequencyLink', 'firmware': 'RFLink Gateway', 'revision': '45', 'node': 'gateway'}
DEBUG:rflink.protocol:got event: {'firmware': 'RFLink Gateway', 'hardware': 'Nodo RadioFrequencyLink', 'revision': '45', 'version': '1.1', 'id': 'rflink'}
rflink                           Nodo RadioFrequencyLink RFLink Gateway V1.1 R45
DEBUG:rflink.protocol:received data: 20;01;Alecto V4;ID=5728;TEMP=00e
DEBUG:rflink.protocol:received data: 0;
DEBUG:rflink.protocol:got packet: 20;01;Alecto V4;ID=5728;TEMP=00e0;
DEBUG:rflink.protocol:decoded packet: {'temperature_unit': '°C', 'temperature': 22.4, 'protocol': 'alecto v4', 'node': 'gateway', 'id': '5728'}
DEBUG:rflink.protocol:got event: {'unit': '°C', 'sensor': 'temperature', 'value': 22.4, 'id': 'alecto v4_5728_temp'}
alecto v4_5728_temp              22.4 °C
DEBUG:rflink.protocol:received data: 20;02;Auriol V2;ID=5E01;TEMP=801
DEBUG:rflink.protocol:received data: d;BAT=OK;
DEBUG:rflink.protocol:got packet: 20;02;Auriol V2;ID=5E01;TEMP=801d;BAT=OK;
DEBUG:rflink.protocol:decoded packet: {'temperature': -2.9, 'protocol': 'auriol v2', 'id': '5e01', 'temperature_unit': '°C', 'battery': 'ok', 'node': 'gateway'}
DEBUG:rflink.protocol:got event: {'unit': None, 'sensor': 'battery', 'value': 'ok', 'id': 'auriol v2_5e01_bat'}
auriol v2_5e01_bat               ok
DEBUG:rflink.protocol:got event: {'unit': '°C', 'sensor': 'temperature', 'value': -2.9, 'id': 'auriol v2_5e01_temp'}
auriol v2_5e01_temp              -2.9 °C
DEBUG:rflink.protocol:received data: 20;03;Alecto V4;ID=5728;TEMP=00e
DEBUG:rflink.protocol:received data: 1;
DEBUG:rflink.protocol:got packet: 20;03;Alecto V4;ID=5728;TEMP=00e1;
DEBUG:rflink.protocol:decoded packet: {'temperature_unit': '°C', 'temperature': 22.5, 'protocol': 'alecto v4', 'node': 'gateway', 'id': '5728'}
DEBUG:rflink.protocol:got event: {'unit': '°C', 'sensor': 'temperature', 'value': 22.5, 'id': 'alecto v4_5728_temp'}
alecto v4_5728_temp              22.5 °C
DEBUG:rflink.protocol:received data: 20;04;Alecto V4;ID=5728;TEMP=00e
DEBUG:rflink.protocol:received data: 1;
DEBUG:rflink.protocol:got packet: 20;04;Alecto V4;ID=5728;TEMP=00e1;
DEBUG:rflink.protocol:decoded packet: {'temperature_unit': '°C', 'temperature': 22.5, 'protocol': 'alecto v4', 'node': 'gateway', 'id': '5728'}
DEBUG:rflink.protocol:got event: {'unit': '°C', 'sensor': 'temperature', 'value': 22.5, 'id': 'alecto v4_5728_temp'}
alecto v4_5728_temp              22.5 °C
DEBUG:rflink.protocol:received data: 20;05;Auriol V2;ID=7501;TEMP=801
DEBUG:rflink.protocol:received data: 0;BAT=LOW;
DEBUG:rflink.protocol:got packet: 20;05;Auriol V2;ID=7501;TEMP=8010;BAT=LOW;
DEBUG:rflink.protocol:decoded packet: {'temperature': -1.6, 'protocol': 'auriol v2', 'id': '7501', 'temperature_unit': '°C', 'battery': 'low', 'node': 'gateway'}
DEBUG:rflink.protocol:got event: {'unit': None, 'sensor': 'battery', 'value': 'low', 'id': 'auriol v2_7501_bat'}
auriol v2_7501_bat               low
DEBUG:rflink.protocol:got event: {'unit': '°C', 'sensor': 'temperature', 'value': -1.6, 'id': 'auriol v2_7501_temp'}
auriol v2_7501_temp              -1.6 °C
DEBUG:rflink.protocol:received data: 20;06;EV1527;ID=045a42;SWITCH=06
DEBUG:rflink.protocol:received data: ;CMD=ON;
DEBUG:rflink.protocol:got packet: 20;06;EV1527;ID=045a42;SWITCH=06;CMD=ON;
DEBUG:rflink.protocol:decoded packet: {'command': 'on', 'switch': '06', 'protocol': 'ev1527', 'node': 'gateway', 'id': '045a42'}
DEBUG:rflink.protocol:got event: {'command': 'on', 'id': 'ev1527_045a42_06'}
`

aequitas commented:
@lmolent then it is some bug specific to HA. Can you add this to your `configuration.yaml` and output some context around one of the Unicode errors:

```
logger:
  default: info
  logs:
    rflink: debug
    homeassistant.components.rflink: debug
```

balloob reviewed (changes requested):
Comment https://github.com/home-assistant/home-assistant/pull/4547#discussion_r94868950 need to be addressed.

lmolent commented:
Sorry @aequitas, it is problem with locale environment pushed to ssh session from my macbook :-)

aequitas commented:
@lmolent good to hear it is resolved, thx for the feedback.

aequitas committed:
Move command logic into separate class.

aequitas committed:
Convert command sending logic to class based pattern instead of using the event bus.

aequitas committed:
Start not using bus for rflink event propagation to platforms.

aequitas committed:
Do not use event bus for all entity types.

aequitas committed:
Fire an event on the bus for every switch incoming rflink command.

aequitas committed:
Resolve lint errors, remove some old code.

aequitas committed:
Known devices no longer need to be registered separately.

aequitas committed:
Log bus events.

aequitas committed:
Event callback is a..... callback.

aequitas committed:
Use full entity id for events.

aequitas committed:
Move event sending to entity.

aequitas committed:
Log incoming events.

aequitas committed:
Make firing events optional inline with rfxtrx.

aequitas committed:
Add foundation for signal repetition.

aequitas committed:
Add signal repetition config and tests.

aequitas committed:
Make plain switchable type explicitly configurable.

aequitas committed:
Enable default entity settings for automatically added entities as well.

aequitas committed:
Prevent default configuration leaking accross entities.

aequitas committed:
Make sure device defaults don't get overwritten by defaults further down.

aequitas committed:
Don't let fast state switching and repetitions turn your house into a disco.

aequitas committed:
Make repetitions more responsive.

aequitas committed:
Disable on/off fallback on dimmables as it currently doesn't play nice with repetitions.

aequitas committed:
Use rflink that allows send_command_ack to be safely cancelled.

aequitas committed:
Reduce duplication and make repeat work for non-ack.

aequitas committed:
Implement reconnection logic.

aequitas committed:
Improve reconnection logic.

aequitas committed:
Also cancel repetitions when entity state is changed due to external command.

aequitas committed:
Update requirements.

aequitas committed:
Fix linting.

aequitas committed:
Fix spelling.

aequitas committed:
Don't lie.

aequitas commented:
@balloob: all significant features are now implemented/tested and your previous comments addressed. I still need to update the website docs. I do not foresee any major changes so you can review the updates.

Quick recap:
- reconnect support added
- signal repetitions added
- no longer uses bus to propagate events internally
- configurable firing of events to bus for button_press event subscription
- some general refactoring and improvement

@lmolent @mjj4791 @KirovAir @Genestealer : Would you mind to test if the latest changes work for you?

@kabturek you had issues before with the tristateinvert, is this still something going on?

aequitas committed:
Fix lint.

mjj4791 commented:
@aequitas I have quickly tested the latest version of the rflink component; it took a while to get the config correct.....

I can confirm the following works for me:
- sensors (including sensors that send multiple values)
- switches/lights
- signal_repetitions
- reconnecting (rflink connected over IP)

One surprise to me (but might be intentional), the entity_id's now contain a space for me: 
"alecto v4_xxxx_hum" and "alecto v4_xxxx_temp" was not expecting the space between alecto and v4.....

aequitas commented:
@mjj4791 thanks for providing such quick feedback. Good to hear everything works for you.

The space in the name should not happen, due to the naming of protocol in Rflink they can contain unserializable characters and I needed to make a translation table: https://github.com/aequitas/python-rflink/blob/master/rflink/parser.py#L269-L284
Downside of this is needs to be maintained with every Rflink release.

I will try to think of a solution where new protocols get added automatically. This will only work if a packet is received by Rflink first and not for devices from configuration.

If I am not mistaking you made some contributions to Rflink which I was not able to incorporate directly. Are all these features now included?

aequitas commented:
@balloob I will be added some modifications to fix @mjj4791 issues and a small feature improvement. No major changes expected.

aequitas committed:
Support for automatically creating protocol translation (fixes spaces in device names).

aequitas committed:
Returned support for dimmable and on/off entity.

aequitas committed:
Duplicate code to fix linting issues with inheritance.

kabturek commented:
@aequitas my problem is with rflink - it reports the value differently. worked around it using iversed switched for now.
will test later today when i get home from vacations. 
great work!

aequitas commented:
@mjj4791 can you try the new updates? Should fix you issue.

aequitas commented:
@kabturek ok, let me know if it is something that can be fixed in code we can always add support for it later.

mjj4791 commented:
@aequitas entity_id's are now without spaces; works for me
It is becoming a cool piece of work!

Question:
I seem to be missing the option to set the unit_of_measurement other than based on the sensor_type....
You currently look it up, based on the sensor_type:

`
config[ATTR_UNIT_OF_MEASUREMENT] = lookup_unit_for_sensor_type(config[CONF_SENSOR_TYPE])
`

If my temperature sensor was sending out, for example, temperature in °F or windspeed in m/s, how would i configure this?


maybe you need to change the following into:
> **if not config[ATTR_UNIT_OF_MEASUREMENT]:**
> .          config[ATTR_UNIT_OF_MEASUREMENT] = lookup_unit_for_sensor_type(config[CONF_SENSOR_TYPE])
> 

and

> .   vol.Optional(CONF_NAME): cv.string,
> .   vol.Required(CONF_SENSOR_TYPE): cv.string,
> .   **vol.Optional(CONF_UNIT_OF_MESUREMENT): cv.string,**
> .   ...

mjj4791 commented:
@aequitas : One more question:
You add newly discovered devices into the 'new_devices_group', which you create on startup of the rflink components. Is it possible for you to not create a group if a group with the same name/id already exists in the config?

This would allow me to place the 'new_devices_group'-s onto a different page in Home Assistant. If I currently try to do that (i.e. define the 'new_rflink_lights' group in my configuration.yaml), I end up with 2 groups with similar name....

Do I do this wrong, or is this not (yet?) possible...?

aequitas committed:
Allow overriding unit of measurement from config.

aequitas commented:
@mjj4791 sorry, I was wrong about the group thing. Currently it is not possible to add devices to an existing group that Rflink has not created. What I think is the common solution for this is set the group to `hidden` using `customize` in the config and then add the new `new_rflink_group` to the view group. But I might be wrong.

Update for configurable unit's has been added.

mjj4791 commented:
@aequitas  I tested the `unit_of _measurement` fix; works for me!

Regarding the groups question, got this working as well:
- define a group in rflink config (for example: `new_rflink_lights` (this group does not/should not exists in your config.yaml as a group)
- add this group into a 'page' (a group with `view: True`)
- in the `customize` config, set the following to make it show up:
` . group.new_rflink_lights:`
` .   hidden: False`
` .  friendly_name: Nieuwe RFLink lampen`

This also works for other default groups (all_devices, all_lights, all_automations, etc.)

balloob commented:
Sorry for the delay, didn't realize you addressed my final comment already 👍  (That's the problem with big PRs…)

aequitas commented:
@balloob no problem, thats why I mentioned you with the status of the PR a few times, so you didn't have to check it in vain :)

Genestealer commented:
Hi,
This is fantastic.
I have this working with the DEV version of HASS.

My EV1527 remote sockets are automatically appearing when the remote is used and HASS can control them just fine. However, my Maclean remote sockets are working, but with errors.

Debug when HASS light button for a Maclean socket is activated (ip removed):

`17-02-07 01:01:51 INFO (MainThread) [homeassistant.components.http] Serving /api/services/homeassistant/turn_on to xx.xx.xx.xx (auth: True)
17-02-07 01:01:51 INFO (MainThread) [homeassistant.core] Bus:Handling <Event call_service[L]: domain=homeassistant, service_call_id=51363536-5, service=turn_on, service_data=entity_id=light.test_switch>
17-02-07 01:01:51 INFO (MainThread) [homeassistant.core] Bus:Handling <Event call_service[L]: domain=light, service_call_id=51363536-6, service=turn_on, service_data=entity_id=['light.test_switch']>
17-02-07 01:01:51 ERROR (MainThread) [homeassistant.core] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "C:\Python\Python35-32\lib\asyncio\tasks.py", line 241, in _step
    result = coro.throw(exc)
  File "C:\Python\Python35-32\lib\site-packages\homeassistant\core.py", line 1052, in _event_to_service_call
    yield from service_handler.func(service_call)
  File "C:\Python\Python35-32\lib\site-packages\homeassistant\components\light\__init__.py", line 242, in async_handle_light_service
    yield from light.async_turn_on(**params)
  File "C:\Python\Python35-32\lib\site-packages\homeassistant\components\rflink.py", line 397, in async_turn_on
    yield from self._async_handle_command("turn_on")
  File "C:\Python\Python35-32\lib\site-packages\homeassistant\components\rflink.py", line 341, in _async_handle_command
    yield from self._async_send_command(cmd, self._signal_repetitions)
  File "C:\Python\Python35-32\lib\site-packages\homeassistant\components\rflink.py", line 367, in _async_send_command
    yield from self._protocol.send_command_ack(self._device_id, cmd)
  File "C:\Users\admin\AppData\Roaming\.homeassistant\deps\rflink\protocol.py", line 158, in send_command_ack
    yield from asyncio.wait_for(self._command_ack.wait(), TIMEOUT)
  File "C:\Python\Python35-32\lib\asyncio\tasks.py", line 385, in wait_for
    yield from waiter
  File "C:\Python\Python35-32\lib\asyncio\futures.py", line 361, in __iter__
    yield self  # This tells Task to wait for completion.
RuntimeError: Task <Task pending coro=<_event_to_service_call() running at C:\Python\Python35-32\lib\site-packages\homeassistant\core.py:1052>> got Future <Future pending> attached to a different loop`

Kind regards,
Richard

lmolent commented:
Hi,
tested newkaku wall switches, some socket switches, temperature sensors and cheap china PIR sensors. Everything looks OK.
Could you pls help me how to use correctly PIR sensors, which is represented by rflink as switch and is sending only ON command when movement is detected. So at this time I must revert switch to OFF state with automate section after few seconds.

aequitas commented:
@Genestealer thanks for the feedback.
- do you have it every time with this specific device or also sometimes or with other devices as well? - Would you be able to test it on a non-windows platform to exclude platform specific bugs. 

I think it has to do with not explicitly passing the event loop down to rflink here: https://github.com/aequitas/python-rflink/blob/master/rflink/protocol.py#L158 
But i'm curious why I have not yet run into this issue before myself as I expect it would pop up more often.

aequitas commented:
@lmolent also thanks for the feedback. For the PIR sensors. What brand/type are they (or more specific what Rflink protocol (http://www.nemcon.nl/blog2/protref))?

Could you maybe run the `rflink -vv` again and capture some debug output for the PIR sensor?

Genestealer commented:
@aequitas - Yes is gives that error for every Maclean TX, but not for the other sockets I have tested.
Just to add extra complexity: HASS runs in a Win 10 32-bit VM with no USB connections to RFLink, so I'm talking to RFLink using telnet to a Wi-Fi NodeMCU ESP8266 running the Arduino "WiFiTelnetToSerial" at 57600 baud. [https://github.com/esp8266/Arduino/blob/master/libraries/ESP8266WiFi/examples/WiFiTelnetToSerial/WiFiTelnetToSerial.ino](url)

The TX and RX pins from the NodeMCU are connected to the RX and TX pins of the Mega via a bi-directional logic level converter.

Perhaps the delay for the data to reach the RFLink via Wi-Fi is greater than your 5 millisecond timeout on https://github.com/aequitas/python-rflink/blob/master/rflink/protocol.py#L158 ?

aequitas commented:
I think you are hitting a specific race condition indeed. Maybe the command for these devices take longer/shorter than others? I will try to push a fix later today.

The timeout should be 5 seconds. Will double-check this also.

lmolent commented:
Hi @aequitas, sorry for delay, these motion sensors are based on chip EV1527, as You can see, only ON command was sent: 

DEBUG:rflink.protocol:received data: 20;06;EV1527;ID=045a42;SWITCH=06
DEBUG:rflink.protocol:received data: ;CMD=ON;
DEBUG:rflink.protocol:got packet: 20;06;EV1527;ID=045a42;SWITCH=06;CMD=ON;
DEBUG:rflink.protocol:decoded packet: {'switch': '06', 'id': '045a42', 'command': 'on', 'node': 'gateway', 'protocol': 'ev1527'}
DEBUG:rflink.protocol:got event: {'id': 'ev1527_045a42_06', 'command': 'on'}
ev1527_045a42_06                 on
DEBUG:rflink.protocol:received data: 20;07;EV1527;ID=045a42;SWITCH=06
DEBUG:rflink.protocol:received data: ;CMD=ON;
DEBUG:rflink.protocol:got packet: 20;07;EV1527;ID=045a42;SWITCH=06;CMD=ON;
DEBUG:rflink.protocol:decoded packet: {'switch': '06', 'id': '045a42', 'command': 'on', 'node': 'gateway', 'protocol': 'ev1527'}
DEBUG:rflink.protocol:got event: {'id': 'ev1527_045a42_06', 'command': 'on'}
ev1527_045a42_06                 on

aequitas commented:
@Lampy09 I will start working on group support.

Sensor type is mandatory for sensors from configuration. I documented this wrong.

Do you only receive the unpack exception on restart/startup? This is a known issue that the initial data is incomplete and can be ignored. I will try to suppress that error.

aequitas commented:
@lmolent no problem. The `rflink -vv` dumps all data received by the Rflink device. If it does not send any data for 'off' there is not much we can do :(. Maybe the rflink device does not properly decode the 'off' commands? Connecting to the rflink manually and turning on debug mode (http://www.nemcon.nl/blog2/development) should give some information on this. You can try the rflink forum/maintainer to have this resolved: http://www.nemcon.nl/blog2/contact

Genestealer commented:
Hi @aequitas there is an issue in the RFLink Gatway firmware interacting with my STATUS RCT-08 remote control sockets. It thinks both ON and OFF button for one channel are different devices. It's RF data is basically any decimal string (between 1 and 9999999) for ON, OFF is simply the 8 less. (E.G. ON: 2006879, OFF: 2006871)

I.E.
Channel ONE, ON Button: `20;53;EV1527;ID=01e9f5;SWITCH=0b;CMD=ON;`
Channel ONE, OFF Button: `20;54;EV1527;ID=01e9f5;SWITCH=03;CMD=ON;`

I also use my own RF MQTT gateway ([https://github.com/Genestealer/ESP8266-433Mhz-Controller-Gateway](https://github.com/Genestealer/ESP8266-433Mhz-Controller-Gateway)), which the RFLink Gateway and your brilliant HASS component understands just fine, but has no way to know the ON and OFF's for the same device.

To make your component even more flexible, would it be possible to have different aliases for the ON and OFF states of a switchable light, the sort of thing used in the MQTT Light “payload_on” & “payload_off”? Or perhaps even go to just comms raw data?

E.G.
```
# Example configuration.yaml entry
light:
  platform: rflink
  devices:
    EV1527_01e9f5:
      name: Living room
      aliasses_on:
        - EV1527_01e9f5_0b_ON
      aliasses_off:
        - EV1527_01e9f5_03_ON
```

If this is too much to ask I can ask RFLink if they could add the STATUS remotes to their supported list over the existing devices.

Many thanks,
Richard

lmolent commented:
Hi, @aequitas "fire_event: true" in "device_defaults:" section e.g:
```
switch:
  platform: rflink
  device_defaults:
    signal_repetitions: 2
    fire_event: true
  devices:
    ev1527_045a42_06:
      name: test1
    ev1527_074842_06:
      name: test2

```
 probably doesn't works. It must be configured in every device section e.g:
```
switch:
  platform: rflink
  device_defaults:
    signal_repetitions: 2
    fire_event: true
  devices:
    ev1527_045a42_06:
      name: test1
      fire_event: true
    ev1527_074842_06:
      name: test2
      fire_event: true
```
Tested with rflink.light and rflink.switch and homeassistant==0.38.4. No button_pressed event is fired.

Thank You.

MacDennis76 commented:
@aequitas : really nice work so far!
I have a RFLink and various KaKu devices and I can help you test/debug if needed.

Tested working for now:
AMT-606
AWMD-250
AWST-8802
And also various Cresta, Auriol & Alecto temperature sensors.
Still need to test motion sensor and various wall plugs.

I also own the ACDB-7000AC doorbell. I had some struggles getting it to work but I found a workaround.

Some debug data:

```
17-05-06 22:46:24 DEBUG (MainThread) [rflink.protocol] received data: 20;07;NewKaku;ID=00c021ca;SWITCH
17-05-06 22:46:24 DEBUG (MainThread) [rflink.protocol] received data: =1;CMD=ALLON;
17-05-06 22:46:24 DEBUG (MainThread) [rflink.protocol] got packet: 20;07;NewKaku;ID=00c021ca;SWITCH=1;CMD=ALLON;
17-05-06 22:46:24 DEBUG (MainThread) [rflink.protocol] decoded packet: {'protocol': 'newkaku', 'switch': '1', 'id': '00c021ca', 'node': 'gateway', 'command': 'allon'}
17-05-06 22:46:24 DEBUG (MainThread) [rflink.protocol] got event: {'command': 'allon', 'id': 'newkaku_00c021ca_1'}
17-05-06 22:46:24 DEBUG (MainThread) [homeassistant.components.rflink] event of type command: {'command': 'allon', 'id': 'newkaku_00c021ca_1'}
```

It sends the 'allon' command but this is a group command. Looking at the RFLink source code group commands only seems to be supported yet in the Light component and not in the Switch component. 

I added the following in my lights configuration:
 
```
platform: rflink
automatic_add: false
device_defaults:
  devices:
  newkaku_00c021ca_1:
    name: Doorbell
    type: toggle
    fire_event: true
```

And I used this automation rule:

```
alias: Doorbell pressed rule
initial_state: 'on'
trigger:
  platform: event
  event_type: button_pressed
  event_data: {'entity_id': 'light.Doorbell', 'state': 'allon'}
action:
  service: logbook.log
  data:
    name: 'Doorbell'
    message: 'was pressed
```

When the doorbell is pressed the automation rule is triggered and I see an entry in the logbook.

Any idea when full support for a doorbell (as a switch) will be implemented? 

Thank you!

aequitas commented:
@MacDennis76 hi, thanks for providing the feedback. I don't have a doorbell so I'm not able to test it myself. Nice to see how Kaku implements this.

I think the doorbell implementation you wrote is as close as it can get. However I would not use the `device_defaults` but just create a `switch` device. Switches support the group commands as well (https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/rflink.py#L390).

There is not really a 'doorbell' concept in HA that i'm aware of (I can't find any other than the Ring component). So I would't not what else to implement. If you have a use case example I can work with I can have a look or if you have a implementation idea feel free to create a PR.
Also does the doorbell send a `alloff` when it is released? Because otherwise the state of the doorbell switch would be constantly on.

I think subscribing to the event is the best solution to implement a doorbell and further logic.

MacDennis76 commented:
@aequitas , thanks for the quick reply!

The doorbell does not send an 'alloff' command.
How would I add the doorbell as a swich? Because I have been unsuccessful so far. Simply moving it from my lights config to my switch config does not work:

```
  newkaku_00c021ca_1:
    name: Doorbell
    fire_event: true
```

I am new to HASS & Python (background is C#) but looking at this line 'allon' will be handled as a group command/event: 
[https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/rflink.py#L134](url)

But the rflink switch component has no logic which initializes the DATA_ENTITY_GROUP_LOOKUP collection, the rflink light component does have this logic and by default a light is part of a group.

When using rflink switch this line is called in the end:
[https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/rflink.py#L145](url)

Is my conclusion correct that group support is still to be implemented in the rflink switch component?

aequitas commented:
@MacDennis76 ah yes, sorry, forgot about that. There is no reason a switch cannot be part of a group. I think I forgot it in the implementation. The event handling is to be overhauled anyways (https://github.com/home-assistant/home-assistant/issues/6767) so I will include it with that change.

Genestealer commented:
@MacDennis76 
Would suggest you try:

### Create the switch, which will fire an event when the doorbell is pressed.
```
switch:
  platform: rflink
    device_defaults:
      fire_event: true
    devices:
      newkaku_00c021ca_1:
        name: Doorbell
```
### Create the automation to react to the event and also set the switch back to the OFF State ready for the next button press.

```
automation:
  alias: "Door Bell Pressed"
  trigger:
    platform: state
    entity_id: switch.doorbell
    state: 'on'
  action:
    - service: notify.google_cloud_notify
      data_template:
        title: 'Doorbell'
        message: 'Someone is at the door!'
    - service: switch.turn_off
       entity_id: switch.doorbell
```

MacDennis76 commented:
@Genestealer Thanks, but if I understand @aequitas last comment correctly a doorbell can not be used as a switch yet. The current code does not call the handle_event method and thus does not fire an event when the doorbell is configured as a switch.

aequitas commented:
@MacDennis76 correct, but if you replace `switch` in @Genestealer example with `light` it should work. Much like the example you provided.

MacDennis76 commented:
@Genestealer @aequitas Yes, I can confirm that automation script does work for a doorbell when used as a light. Looking forward to #6767 . If you need help testing any devices just let me know.

aequitas commented:
@MacDennis76 I will keep progress in that issue (and the subsequent pullrequest). If you subscribe (notifications) to the issue you will be notified when there are changes.

Genestealer commented:
As an alternative you could learn the doorbell has a custom RF switch within the Arduino memory which would overcome your issue.

jeroenterheerdt commented:
guys - experimenting with RFLink now. I am a little familiar with HA (did some development myself) but do not get how this RFLINK component is meant to work. I see in the logs lots of 'device_id not known, adding new device'. Afaik nothing is happening afterwards. What is supposed to happen? I do not see my devices show up in the UI nor do I see anything happening in the configuration.yaml...

jeroenterheerdt commented:
@aequitas any idea in working with two rflinks connected to the same HA? I have the 433 one but want to add the 867 as well. Not sure how that will work with HA...

Genestealer commented:
Have you tried putting both the 433Mhz and 868Mhz transmitters in parallel on the Arduino Mega? You may need to parallel up two level shifters if you have problems but I cannot see why having them in parallel wouldn't work.

aequitas commented:
@jeroenterheerdt devices should appear on the homepage when detected. If not it might be your configuration hides them by default. 

Alternatively add the device id's manually to the config as described in the documentation and see if that works. 

I have not worked on supporting double receivers. The current implementation of python-rflink supports only one device at a time. I don't know if HA allows you to configure a component twice, never tried this.

jeroenterheerdt commented:
@genestealer, how would putting a parallel work? @aequitas, I am pretty certain that HA will not allow double activation of the same component.

Genestealer commented:
@jeroenterheerdt, connect both transmitters Tx pins to MEGA tx3/Pin 14 (TX Data). then the data should be transmitted by both transmitters.

jeroenterheerdt commented:
@genestealer do you have a sample of a double transmitter / receiver?

Genestealer commented:
@jeroenterheerdt No sorry, I would suggest you give it a go and see.

If not, back in Aug 2016 the Stuntteam did say "RFlink will be able to use both frequencies in an upcoming version."

jeroenterheerdt commented:
@genestealer I have talked to the Nodoship guys and they told me it cannot be done this way. So, maybe we just need to solve this software side. Is there any way we could make your lib able to handle two rflinks?

Genestealer commented:
I wouldn't necessarily believe that. Do you have both transmitters to try?

aequitas commented:
This could be implemented in either https://github.com/aequitas/python-rflink or the HA component. Currently all connection logic is handled in the HA component (https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/rflink.py#L179). So that means either moving all that logic to python-rflink and simplifying the HA component or extend the HA component to allow creating and handling multiple connections. Both are pretty big tasks which I'm unable to put on my roadmap atm, but if you want to give a shot at implementing it yourself I can assist if needed.

jeroenterheerdt commented:
@genestealer, no I do not and I also do not have the other components you mentioned. I had a look at the support forum for rflink and that give me not a lot of hope. The same question is answered there with a firm 'cannot be done'.
